### ISSUE TITLE
Allow items after statements by default

### ISSUE THREAD

**Issue by tikue** [Manishearth/rust-clippy#722]
Putting items after statements puts the actual logic of the `fn` front and center. This is a style thing that I'd imagine many people are strongly opinionated about.

As one data point, here is code from @nikomatsakis's Rayon [blog post](http://smallcultfollowing.com/babysteps/blog/2016/02/25/parallel-iterators-part-2-producers/):

```rust
impl IndexedParallelIterator for MapParIter<ITER, MAP_OP, RET>
    where ITER: ParallelIterator,
          MAP_OP: Fn(ITER::Item) -> RET + Sync
{
    fn with_producer<CB>(self, callback: CB)
        where CB: ProducerCallback<RET>
    {
        let my_callback = MyCallback { // defined below
            callback: callback,
            map_op: &self.map_op,
        };

        self.base.with_producer(my_callback);

        struct MyCallback<'m, MAP_OP, CB> {
            //          ^~
            //
            // This is that same lifetime `'m` we had trouble with
            // in the previous example: but now it only has to be
            // named from *inside* `with_producer`, so we have no
            // problems.

            callback: CB,
            map_op: &'m MAP_OP
        }

        impl<'m, ITEM, MAP_OP, CB> ProducerCallback<ITEM> for MyCallback<'m, MAP_OP, CB>
            where /* omitted for "brevity" :) */
        {
            type Output = (); // return type of `callback`

            // The method that `self.base` will call with the
            // base producer:
            fn callback<P>(self, base_producer: P)
                where P: Producer<Item=ITEM>
            {
                // Wrap the base producer in a MapProducer.
                let map_producer = MapProducer {
                   base: base_producer,
                   map_op: self.map_op,
                };

                // Finally, callback the original callback,
                // giving them out `map_producer`.
                self.callback.callback(map_producer);
            }
        }
    }
}
```

**Comment by mcarton** [comment]
I vote for. I’ve already been annoyed by that lint, the inner function was declared near where it was used, I though this was better that at the top of the outer function.

**Comment by llogiq** [comment]
:+1: we're already quite opinionated as is and having inner items at the start of a function doesn't improve legibility in all cases.

**Comment by tikue** [comment]
The example in the docs for this lint is compelling, though; it may be worthwhile to warn by default against the more specific case of `item` => `statement` => `item`.

**Comment by llogiq** [comment]
I most often see this in OS-handling code where you have operations branched out for e.g. Windows and Unixes. You'd see `item item statement item item statement` where each item had a `#[cfg(..)]` pretended. Your revised rule would still catch this.

**Comment by tikue** [comment]
Ah, fair enough! Yeah, in that case, I think the likelihood of a "false positive" for this lint merits allow by default.

**Comment by oli-obk** [comment]
Aren't cfgs eliminated in hir?

**Comment by Manishearth** [comment]
cfgs are eliminated quite early. Lints don't get to see pre-cfg stuff.

**Comment by llogiq** [comment]
That still leaves `item statement item statement.
